Add Mine Game Arcade Machine

Joe-Derp opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Adds a very well known game involving locating and clearing a tile grid without hitting any mines as an arcade machine.

Inspiration taken from https://github.com/yogstation13/Yogstation/pull/15629 / https://github.com/yogstation13/Yogstation/pull/4996 (and the original 13 PR before it was ported to Yogs13)

## Why / Balance
I believe this would help to complement the open-endedness of the Block Game Arcade and the annoying, fixed chance Space Villain Arcade which feels like it's always down to luck. This new arcade machine game would be an ideal balance of skill and luck, and supports a good source of fun for station staff like tiders (keeping them from fiddling with electrical wires).
The luck-skill balance would also make it a good candidate for prize vending in the future (the prize code is unorganized and heavily linked into the Space Villain arcade currently and may have more balancing ramifications, so I'd likely only propose it as another later PR if this were first accepted). 

## Technical details
Adds a new arcade machine, primarily based off the code/resource structuring of the already existing Space Villain and Block Game arcade machine.
In that regard, similarly uses same BoundUI net messages to transfer data only to active players (# of which is unrestricted; multiple people can help clear one board at once) of an arcade machine.
Minimal game visual updating code is present on client for better responsiveness, while all primary game logic is done on server before sending visual state to client for rendering.
Client visuals are made with grid of controls, which could use optimizing either with tricks for better ui performance or entirely different rendering structure in the future to support larger board sizes.
**(edit)**
Code structuring and networking updated to hopefully be based off more modern practices, uses an auto networked shared component (in place of BUI messages to update client state) and shared entitysystem, with predicted messages. 

## Media
https://github.com/user-attachments/assets/5af646e6-49a5-4b4a-b3f4-e80955351293

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
N/A

**Changelog**
:cl: joederpington
- add: Added Mine Game Arcade Machine

Joe-Derp committed:
Add Mine Clearing Arcade Machine

Commit:
Fix mine game machine sprite timing

Commit:
Fix missing ad & improperly removed component

Warpzoned commented:
> will it be emaggable?

lmao if it is I hope that once you click a mine the cab blows up

slarticodefast reviewed (changes requested):
Fun idea!
I assume this copies a lot of ancient code from the other arcade games, which is bad.
The UI should be predicted.

Also some other comments of stuff I saw with a quick look, so this is not a full review.

slarticodefast commented on Content.Client/Arcade/MineGameMenu.cs:
```diff
@@ -0,0 +1,331 @@
+/// <summary>
+/// Client UI Window for 'Mine Game' Arcade Machine
+/// </summary>
+public sealed class MineGameMenu : DefaultWindow
```
Oh, and you can make windows with xaml, no need to manually create every single button and label in C#
And use a FancyWindo.

Joe-Derp replied:
I'm aware of being able to make windows with XAML. I felt like there weren't too many UI elements and most of the UI is more code-heavy than simply being static layout. I can try to convert some of it to use the XAML JIT compiler if you'd like.

Joe-Derp commented on Resources/Textures/Interface/Misc/arcade_minegametile.rsi/meta.json:
```diff
@@ -0,0 +1,53 @@
+        "y": 16
+    },
+    "license": "CC-BY-SA-3.0",
+    "copyright": "Made by Joe-Derp",
```
They are made from scratch, I was expecting it would be more noticeable as coder art and not taken from any previous game sprites which may have licensing conflicts. The license is applied myself.

Joe-Derp commented on Resources/Textures/Structures/Machines/arcade.rsi/screen_minegame.png:
This seemed like it was standard for all of the arcade machines. From what I understood, it requires the same size for all four sprite directions, but the arcade machine screens are only visible on the frontward facing sprite. Thus, all of the arcade sprites have this empty space. If there's a way to just get rid of it, please let me know and I can correct it.

Joe-Derp commented:
> will it be emaggable?



> LOVE this. Instant massive win for bored prisoners.
> 
> Emagging it should cause hitting a mine to actually blow the machine up.



> Emag should explode, hacking it should make it do something silly though

I would like to add an emagging functionality since those were present in the ss13 inspiration although I saw that 'emag interactions' were in the freezes/restrictions list and I didn't want to potentially muddy this PR with it.

slarticodefast commented on Content.Server/Arcade/MineGame/MineGame.cs:
```diff
@@ -0,0 +1,292 @@
+/// <summary>
+/// Handles game logic of the Mine Game.
+/// </summary>
+public sealed partial class MineGame
```
The problem is that predicting the game state will be more complicated this way. You might be able to turn it into a data definition, but this has to be serializable and networked somehow so that it is persistent and you can predict it.

Joe-Derp replied:
not gonna worry about this for now^ but I have some concerns too about trying to auto network it [discord message link](https://discord.com/channels/310555209753690112/310555209753690112/1424107035573882911)

Commit:
Refactor MineGame client window to use xaml

Commit:
Minor server refactor

Commit:
Small xaml ui fixes

Commit:
Improve shared naming/organization

Commit:
fancy window

Commit:
ECS Refactor

Commit:
Fix issue from potentially missing component field

Commit:
Better networking code + other refactors/cleaning

Joe-Derp committed:
Dial back the DataFields

Was hoping for the option to have a mapped-in / yaml defined game state for funsies but the tests are angry

Joe-Derp committed:
Merge branch 'master' into minegame

Joe-Derp committed:
Fix JSON formatting in arcade.rsi/meta.json
